OpenTelemetry API Tracing.

sfc-gh-rjenkins opened the pull request:
The PR introduces a new distributed tracing implementation called an `OTELSpan` which attempts to provide greater compatibility with the W3C Trace-Context specification https://www.w3.org/TR/trace-context and the OpenTelemetry specification and API https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md.

Unit testing was added for this new implementation.

[Design Document](https://docs.google.com/document/d/1CLlQJ-evP0PGVLwoyC8eF_-jpU8v8NbsR92HOzh2YhM/edit#)

N.B. This provides the implementation of Open Telemetry compliant traces but doesn't yet update the call sites to use this new tracing format. A follow up PR will handle those additional changes.

# Code-Reviewer Section

The general guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [ ] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [ ] Every function/class/actor that was touched is reasonably well documented.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `master` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)

sfc-gh-ljoswiak commented:
The design doc is not accessible outside Snowflake.

sfc-gh-rjenkins commented on flow/Tracing.actor.cpp:
```diff
@@ -403,3 +549,60 @@ Span::~Span() {
+	end = o.end;
+	location = o.location;
+	links = std::move(o.links);
+	events = std::move(o.events);
```
TODO: Add Status and Kind.

PierreZ commented:
Awesome news :smile: Do you have an idea on which release it might be released? I would love to see this in `7.0`

sfc-gh-rjenkins commented:
> The design doc is not accessible outside Snowflake.

👍 I wanted you and Markus to have a look before opening it up.

sfc-gh-ljoswiak reviewed (commented):
I don't have strong feelings on whether we should continue to support the old OpenTracing implementation. I think it could be helpful for anyone that hasn't updated their infrastructure yet, as there seems to be a lot of tooling that is compatible with OpenTracing but not necessarily OpenTelemetry. But it also complicates the implementation it seems. If the common functionality could easily be factored out I don't see much additional downside.

The design doc mentions

> Additionally, if we support multiple wire formats the OpenTelemetry receiver will need to be modified with a configuration parameter to choose which wireformat to expect, or optionally we could add a header with version and schema information so the receiver can detect this dynamically.

Would this add additional overhead to tracing?

sfc-gh-rjenkins commented on flow/Tracing.h:
```diff
@@ -106,6 +107,170 @@ struct Span {
+// 8. An optional list of timestamped Events.
+// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events
+
+enum class SpanKind : uint8_t { CLIENT = 0, SERVER = 1, PRODUCER = 2, CONSUMER = 3, INTERNAL = 4 };
```
Should we namespace these enums?

sfc-gh-rjenkins commented:
Reviewed with @sfc-gh-mpilman last Friday, we're going to move forward with this design but we have to address a few issues. 

1. Resolving whether 16 byte wide traceid should compose 8 bytes for span id.
2. Use Arena in implementation for vectors
3. Move OTELSPan to a class
4. Implement allocator for map so we can use with arena.

I will be working on these changes this week.

sfc-gh-rjenkins commented:
> Awesome news 😄 Do you have an idea on which release it might be released? I would love to see this in `7.0`

I think it's currently targeted for 7.1 but will double check.

sfc-gh-rjenkins commented:
@sfc-gh-ljoswiak, @sfc-gh-mpilman, and @sfc-gh-anoyes. This is now ready for review again. Please keep an eye on the memory management and verify the use of Arenas is safe. Some notes about the latest changes.

**Passing const reference.** 

I've changed most of the functions in Tracing.h to accept const references rather than pass-by-value. I believe all these spans are generally created on the stack and are short lived, being serialized by the tracer when the stack unwinds. As they're not being dereferenced on the heap afaik, I figured this would avoid a copy and be a little faster. If you'd prefer they be passed by value lmk. 

**Fluent api.**

I added a fluent api on the addLinks/addEvents/AddAttributes API, so you can create a span and do method chaining like... 
`auto s = OTELspan("foo"_loc).addAttribute("foo", "bar").addLink(linkedContext);`. Some people like this, others strongly dislike because you can have an error and it's difficult to tell with method and parameters were the cause. I don't know what people on this team prefer so lmk if you like it or you want it removed.

**Prefer StringRef/KeyValueRefs or const std::string& or std::string?**

The previous implementation took StringRefs for the tags map. For the AddEvents/AddAttributes method I actually implemented overloads that could accept either `const std::string&` or a `StringRef`. However, Apple CI threw errors about ambiguity, I suspect due to implicit conversion of `const std::string&` to a StringRef https://github.com/apple/foundationdb/blob/main/flow/Arena.h#L457? Anyway I commented those methods out which took a const std::string& rather than a StringRef for now, as I wanted to discuss this issue here. In summary, I'm not clear on what we *should* prefer.

It would seem to me that from a defensive standpoint if we want to use Arena, we should accept std::strings and construct our own StringRefs using the OTELSpan's own Arena, or at least link the two Arenas. The older `Span` implementation accepted and stored StringRefs for tags in a std::unordered_map, without an Arena. Afaik from the docs https://github.com/apple/foundationdb/tree/main/flow#arenas, that data *could be deallocated and those `StringRefs` made invalid. However, seeing as these are all short lived, perhaps it is OK to not be so defensive here and trust the user on the lifetime of the accepted StringRefs. I asked @sfc-gh-anoyes about best practice here, but he explained there isn't a hard and fast rule.

**Check TODOs in Tracing.h**

Double check the TODOs, I've left them because I wanted feedback. Most of them are in Tracing.h and revolve around the Arena and safety preferences. 

**Follow up work**
I suggest the next steps get done in a separate PR.

1. Decide on keeping both implementations or not?
2. Update call sites

If we want both, we'll need a new Knob to select the Span type configured, and then we'll need to update the call sites with conditionals to create the right spans. We could probably tuck that away with some indirection, which may add some overhead.

If we want to remove the opentracing.io Span, we can do that in a follow up PR, I'll rename OTELSpan->Span and then update the call sites.

sfc-gh-mpilman reviewed (approved):
This looks good to me. It might be worth adding a comment in the PR description that this is just the implementation of the framework and it doesn't yet add OpenTelemetry support to FDB.

sfc-gh-rjenkins commented:
> This looks good to me. It might be worth adding a comment in the PR description that this is just the implementation of the framework and it doesn't yet add OpenTelemetry support to FDB.

OK, will do. I'm also going to delete those helper methods I commented out that take `const std::string&` if you are OK with the use of StringRef in all those.

sfc-gh-ljoswiak reviewed (commented):
I left mostly nits. In regards to whether to accept StringRef/std::string/both, I think accepting just StringRefs, as we currently do, works. Since the serialization happens on span destruction, I think the data should still be valid.

sfc-gh-mpilman commented on flow/Tracing.actor.cpp:
```diff
@@ -242,6 +283,67 @@ struct UDPTracer : public ITracer {
+			request.write_byte(static_cast<uint8_t>(size) | 0b10000000);
+		} else {
+			// TODO: Add support for longer maps if necessary.
+			ASSERT(false);
```
Alternatively you can use `ASSERT_WE_THINK(false)`. This will only cause a crash in simulation but not in prod. Additionally I would still log something (just to make debugging easier should we ever run into this issue).

sfc-gh-anoyes commented on flow/Tracing.actor.cpp:
```diff
@@ -151,7 +180,7 @@ ACTOR Future<Void> traceLog(int* pendingMessages, bool* sendError) {
 
 struct UDPTracer : public ITracer {
-protected:
+public:
```
```suggestion
```

sfc-gh-anoyes commented on flow/Tracing.actor.cpp:
```diff
@@ -403,3 +533,350 @@ Span::~Span() {
 	}
 }
+
+OTELSpan& OTELSpan::operator=(OTELSpan&& o) {
```
Serializing while moving seems strange to me. Do we need a move constructor at all? @sfc-gh-ljoswiak @sfc-gh-mpilman

sfc-gh-anoyes replied:
Also it appears that `OTELSpan& operator=(OTELSpan&& o)` and `OTELSpan(OTELSpan&& o)` have different behavior, which seems strange

sfc-gh-ljoswiak replied:
You mean the behavior is different such that the assignment operator serializes the span whereas the move constructor doesn't? I guess it's a little strange. We use this behavior in loops when we want to make sure the existing span gets serialized before assigning a new one. It takes some work off the implementor because then they don't have to remember to call `span.trace()` or some such function.

I don't know if the move constructor is required. I guess you could remove it and see if anything breaks?
